PathPlanning/DStarLite/DStarLite.py

```diff
@@ -0,0 +1,278 @@
+Implemented maintaining similarity with the pseudocode for understanding
+"""
+from __future__ import annotations
+import math
+import matplotlib.pyplot as plt
+import random
+
+show_animation = True
+pause_time = 0.1
+p_create_random_obstacle = 0
+
+
+class Node:
+    def __init__(self, x: int = 0, y: int = 0, cost: float = 0.0):
+        self.x = x
+        self.y = y
+        self.cost = cost
+
+
+def add_coordinates(node1: Node, node2: Node):
+    new_node = Node()
+    new_node.x = node1.x + node2.x
+    new_node.y = node1.y + node2.y
+    new_node.cost = node1.cost + node2.cost
+    return new_node
+
+
+def compare_coordinates(node1: Node, node2: Node):
+    return node1.x == node2.x and node1.y == node2.y
+
+
+class DStarLite:
+
+    motions = [
+        Node(1, 0, 1),
+        Node(0, 1, 1),
+        Node(-1, 0, 1),
+        Node(0, -1, 1)
+    ]
+
+    def __init__(self, x_max: int, y_max: int, obstacles: list,
+                 spoofed_obstacles: list = list()):
+        self.x_max = x_max
+        self.y_max = y_max
+        self.obstacles = obstacles
+        self.spoofed_obstacles = spoofed_obstacles
+        self.start = Node(0, 0)
+        self.goal = Node(0, 0)
+        self.U = list()
+        self.km = 0
+        self.kold = 0
+        self.rhs = list()
+        self.g = list()
+        self.detected_obstacles = list()
+        if show_animation:
+            self.detected_obstacles_for_plotting_x = list()
+            self.detected_obstacles_for_plotting_y = list()
+
+    def create_grid(self, val: float):
+        grid = list()
+        for _ in range(0, self.x_max):
+            grid_row = list()
+            for _ in range(0, self.y_max):
+                grid_row.append(val)
+            grid.append(grid_row)
+        return grid
+
+    def is_obstacle(self, node: Node):
+        return any([compare_coordinates(node, obstacle)
+                    for obstacle in self.obstacles]) or \
+               any([compare_coordinates(node, obstacle)
+                    for obstacle in self.detected_obstacles])
+
+    def c(self, node1: Node, node2: Node):
+        if self.is_obstacle(node1) or self.is_obstacle(node2):
+            # Attempting to move from or to an obstacle
+            return math.inf
+        new_node = Node(node1.x-node2.x, node1.y-node2.y)
+        detected_motion = list(filter(lambda motion:
+                                      compare_coordinates(motion, new_node),
+                                      self.motions))
+        return detected_motion[0].cost
+
+    def h(self, s: Node):
+        # Modify if allowing diagonal motion
+        return abs(self.start.x - s.x) + abs(self.start.y - s.y)
+
+    def calculate_key(self, s: Node):
+        return (min(self.g[s.x][s.y], self.rhs[s.x][s.y]) + self.h(s)
+                + self.km, min(self.g[s.x][s.y], self.rhs[s.x][s.y]))
+
+    def is_valid(self, node: Node):
+        if node.x >= 0 and node.y >= 0 and node.x < self.x_max and \
```

Comment:
`if 0 <= node.x < self.x_max and 0 <= node.y < self.x_max:`
is better?

Reply:
Yes, that works.